### PR TITLE
Memory use in `bin`

### LINKED ISSUE #1491: High memory consumption (and bad performance) in `bin` with many existing bins
The current implementation of `bin` use `O(M*N)` memory for temporary index handling, when "rebinning" `M` input bins into `N` output bins. This gets problematic in real applications, e.g., for neutron event data with 1 M pixels (bins), rebinned into 1000 bins for display, `M*N = 1e9`. Multiple 4 or 8 byte indices are required, i.e., we quickly exceed 10 GByte.

This is not just a memory issue but also reduced performance, I think.

Provided that the input bins being rebinned have a bin coordinate (as opposed to just an event coord), we actually know which input bins overlap with which output bins. In reality there are thus only `O(M+N)` non-zero entries in the sub-size matrix (and derived parameters).

https://github.com/scipp/scipp/tree/bin-memory-use outlines the basic idea of the refactor.

Note that if the input does not have bins but groups, when binning those groups the same logic applies (strictly speaking there are even slightly fewer non-zero entries).

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #1491. See there for further details.

Marked as draft for now. I think there is more cleanup to do, and I suggest doing a walk-through on Monday, so we can discuss whether this is the correct approach, and decide how to proceed.

Before (killed because of too much memory consumption, also very slow when memory consumption increases):
```
BM_rebin_outer/10/100000          155244789 ns    155244007 ns            4 events=100k items_per_second=644.147k/s xbins=10 ybins=4
BM_rebin_outer/100/100000         422588231 ns    422575119 ns            2 events=100k items_per_second=236.644k/s xbins=100 ybins=4
BM_rebin_outer/1000/100000       2850374029 ns   2850238446 ns            1 events=100k items_per_second=35.0848k/s xbins=1000 ybins=4
BM_rebin_outer/10000/100000      39286912387 ns   39277293308 ns            1 events=100k items_per_second=2.546k/s xbins=10k ybins=4
Killed
```

After:
```
BM_rebin_outer/10/100000           97543629 ns     97536050 ns            7 events=100k items_per_second=1025.26k/s xbins=10 ybins=4
BM_rebin_outer/100/100000          97835014 ns     97826362 ns            7 events=100k items_per_second=1022.22k/s xbins=100 ybins=4
BM_rebin_outer/1000/100000         99146151 ns     99130976 ns            7 events=100k items_per_second=1008.77k/s xbins=1000 ybins=4
BM_rebin_outer/10000/100000       101061036 ns    101051331 ns            7 events=100k items_per_second=989.596k/s xbins=10k ybins=4
BM_rebin_outer/100000/100000      187891911 ns    185731232 ns            4 events=100k items_per_second=538.412k/s xbins=100k ybins=4
BM_rebin_outer/1000000/100000    1638166431 ns   1372812094 ns            1 events=100k items_per_second=72.8432k/s xbins=1000k ybins=4
BM_rebin_outer/10/1000000         100683026 ns    100680133 ns            6 events=1000k items_per_second=9.93245M/s xbins=10 ybins=4
BM_rebin_outer/100/1000000        110754467 ns    110743015 ns            7 events=1000k items_per_second=9.02991M/s xbins=100 ybins=4
BM_rebin_outer/1000/1000000       100365934 ns    100365495 ns            7 events=1000k items_per_second=9.96358M/s xbins=1000 ybins=4
BM_rebin_outer/10000/1000000      106257799 ns    106255060 ns            7 events=1000k items_per_second=9.41132M/s xbins=10k ybins=4
BM_rebin_outer/100000/1000000     175251060 ns    175237164 ns            4 events=1000k items_per_second=5.70655M/s xbins=100k ybins=4
BM_rebin_outer/1000000/1000000   1393924760 ns   1259861949 ns            1 events=1000k items_per_second=793.738k/s xbins=1000k ybins=4
BM_rebin_outer/10/10000000        167359746 ns    167345478 ns            4 events=10M items_per_second=59.7566M/s xbins=10 ybins=4
BM_rebin_outer/100/10000000       166611504 ns    166605160 ns            4 events=10M items_per_second=60.0222M/s xbins=100 ybins=4
BM_rebin_outer/1000/10000000      165337964 ns    165331914 ns            4 events=10M items_per_second=60.4844M/s xbins=1000 ybins=4
BM_rebin_outer/10000/10000000     167763252 ns    167759568 ns            4 events=10M items_per_second=59.6091M/s xbins=10k ybins=4
BM_rebin_outer/100000/10000000    232244943 ns    232236262 ns            3 events=10M items_per_second=43.0596M/s xbins=100k ybins=4
BM_rebin_outer/1000000/10000000  1266291849 ns   1266044067 ns            1 events=10M items_per_second=7.89862M/s xbins=1000k ybins=4
BM_rebin_outer/10/100000000       703813896 ns    703766326 ns            1 events=100M items_per_second=142.093M/s xbins=10 ybins=4
BM_rebin_outer/100/100000000      699346067 ns    699308890 ns            1 events=100M items_per_second=142.998M/s xbins=100 ybins=4
BM_rebin_outer/1000/100000000     763824519 ns    759803177 ns            1 events=100M items_per_second=131.613M/s xbins=1000 ybins=4
BM_rebin_outer/10000/100000000    710722828 ns    709024143 ns            1 events=100M items_per_second=141.039M/s xbins=10k ybins=4
BM_rebin_outer/100000/100000000   976721186 ns    970323061 ns            1 events=100M items_per_second=103.058M/s xbins=100k ybins=4
BM_rebin_outer/1000000/100000000 1880668217 ns   1849596394 ns            1 events=100M items_per_second=54.0658M/s xbins=1000k ybins=4
```

As a more practical example: As a side effect of this a plot for a WISH file with 6.5 GByte of event data and 200k pixels now opens in `3 seconds` instead of `5 seconds`.

### CODE REVIEW

**Comment by Matthew-Andrew** [comment]
Whilst I agree that this changes more files than would be ideal to resolve the memory issue I can't think of a better approach. 

My only ideas for an alternate approach are either:
1) Ignore the input bins on rebinning and just copy all the events into a new buffer and treat it like creating a new binned object. As discussed on monday however this would require a copy in addition to whatever copies are required during re-ordering. It would allow you to skip the large intermediate matrix however so its not completely clear this would be worse. The fuzzy sort in place would help a lot here if possible.
2) Create a Variable type which could handle sparse data with an initial offset. This would gave the advantage of fitting more naturally into the core of scipp and may well have use elsewhere. I remember you saying that this wouldn't solve all the problems though. Could you elaborate on this?

**Comment by SimonHeybrock** [comment]
> * Ignore the input bins on rebinning and just copy all the events into a new buffer and treat it like creating a new binned object. As discussed on monday however this would require a copy in addition to whatever copies are required during re-ordering. It would allow you to skip the large intermediate matrix however so its not completely clear this would be worse. The fuzzy sort in place would help a lot here if possible.

You are right that it would be better in certain cases (few events and many bins), but we also need to optimize for the "many events" case, so one extra copy is not affordable, I think.

> * Create a Variable type which could handle sparse data with an initial offset. This would gave the advantage of fitting more naturally into the core of scipp and may well have use elsewhere. I remember you saying that this wouldn't solve all the problems though. Could you elaborate on this?

I think a generic sparse variable type would work for about two thirds of the operations we need. There are some specifics though which are particular to this application, e.g., during/after `cumsum` we do not want do fill in new areas of the matrix that were previously zero (not stored), even though they would conceptually contain (constant along dim) non-zero values now. That is, we would still need to implement some special functions in addition to the sparse variable type. I think this is something we may switch to eventually, if we ever have concrete requirements to support sparse-matrix type variables.

**Comment by Matthew-Andrew** [comment]
Thanks for answering my outstanding questions @SimonHeybrock. 
I'm happy with this implementation if others are, given the size and the complexity is probably worth at least one other person giving it a look through.

File: core/test/subbin_sizes_test.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+TEST_F(SubbinSizesTest, minus) {
+  SubbinSizes x(2, {1, 2, 3});
+  EXPECT_EQ(x - SubbinSizes({3, {-4, -5, -6}}), SubbinSizes(2, {1, 6, 8, 6}));
+  EXPECT_EQ(x - SubbinSizes({3, {-1}}), SubbinSizes(2, {1, 3, 3}));
+  EXPECT_EQ(x - SubbinSizes({1, {-1, -2}}), SubbinSizes(1, {1, 3, 2, 3}));
+  EXPECT_EQ(x - SubbinSizes({6, {-1}}), SubbinSizes(2, {1, 2, 3, 0, 1}));
+  EXPECT_EQ(x - SubbinSizes({0, {-1}}), SubbinSizes(0, {1, 0, 1, 2, 3}));
+}
```
Comment (OwenArnold) [inline core/test/subbin_sizes_test.cpp:107]: 👍

File: core/subbin_sizes.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#include "scipp/core/subbin_sizes.h"
+
+namespace scipp::core {
+
+SubbinSizes::SubbinSizes(const scipp::index offset, container_type &&sizes)
+    : m_offset(offset), m_sizes(std::move(sizes)) {}
+
+SubbinSizes::SubbinSizes(const scipp::index value)
+    : m_offset(0), m_sizes({value}) {}
+
```
Comment (OwenArnold) [inline core/subbin_sizes.cpp:19]: This is an internal type so, unlikely to get exposed to misuse. However, did you consider how this behaves with bad inputs negative offsets etc. I suspect checking would hit performance even if it was considered worthwhile.

File: core/test/element_bin_util_test.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    EXPECT_EQ(index, expected);
+  };
+  check(-0.1, 2);
+  check(0.0, 2);
+  check(1.0, 2);
+  check(1.5, 3);
+  check(2.0, 3);
+  check(2.5, 4);
+  check(3.0, 4);
+}
```
Comment (OwenArnold) [inline core/test/element_bin_util_test.cpp:44]: Missing tests for `subbin_sizes_exclusive_scan` ?

**Review by OwenArnold — COMMENTED** [review]
Generally looks good. I didn't identify any issues with the code.

File: core/test/subbin_sizes_test.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+using namespace scipp::core;
+
+class SubbinSizesTest : public ::testing::Test {};
+
+TEST_F(SubbinSizesTest, comparison) {
+  EXPECT_TRUE(SubbinSizes(2, {1, 2, 3}) == SubbinSizes(2, {1, 2, 3}));
+  EXPECT_FALSE(SubbinSizes(2, {1, 2, 3}) == SubbinSizes(2, {1, 2}));
+  EXPECT_FALSE(SubbinSizes(2, {1, 2, 3}) == SubbinSizes(2, {1, 2, 4}));
+  EXPECT_FALSE(SubbinSizes(2, {1, 2, 3}) == SubbinSizes(3, {1, 2, 3}));
+  // If this was a generic type we should probably return true here, but as it
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/test/subbin_sizes_test.cpp:17]: Can you elaborate? `==` returns `false` in the following two lines.

File: core/subbin_sizes.cpp (hunk 2 above)
Comment (SimonHeybrock) [reply core/subbin_sizes.cpp:19]: I think performance should not be affected much, I can add a check.

File: core/test/subbin_sizes_test.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (OwenArnold) [reply core/test/subbin_sizes_test.cpp:17]: Sorry my bad. Too much time scanning through code on github this morning.

File: core/test/element_bin_util_test.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply core/test/element_bin_util_test.cpp:44]: Adding one in `variable/test/bin_util_test.cpp`. Testing the kernel in isolation is actually not so useful in this case.

Comment (SimonHeybrock) [reply core/test/element_bin_util_test.cpp:44]: > Adding one in `variable/test/bin_util_test.cpp`. Testing the kernel in isolation is actually not so useful in this case.

The question is how useful it is to test such an implementation detail. Should maybe rather use the time to write more tests for `bin` itself, so even if we refactor the implementation later we do not waste the development time?

Comment (SimonHeybrock) [reply core/test/element_bin_util_test.cpp]: I moved those function to a different file and namespace to make clear that they are implementation details.

**Review by OwenArnold — APPROVED** [review]

File: variable/bin_detail.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+/// Index of the bin (given by `edges`) containing a coord value.
+///
+/// 0 if the coord is less than the first edge, nbin-1 if greater or equal last
+/// edge. Assumes both `edges` and `coord` are sorted.
+Variable begin_edge(const VariableConstView &coord,
+                    const VariableConstView &edges) {
+  auto indices = makeVariable<scipp::index>(coord.dims());
+  const auto dim = edges.dims().inner();
+  Variable bin(indices.slice({dim, 0}));
```
Comment (jl-wynen) [inline variable/bin_detail.cpp:25]: What if dim is empty? (Same in `end_edge`)

File: variable/string.cpp [hunk 6]
```diff
@@ -69,6 +69,7 @@ auto apply(const DType dtype, Args &&... args) {
       std::tuple<double, float, int64_t, int32_t, std::string, bool,
                  scipp::core::time_point, Eigen::Vector3d, Eigen::Matrix3d,
                  Variable, bucket<Variable>, bucket<VariableConstView>,
+                 bucket<Variable>, bucket<VariableConstView>,
```
Comment (jl-wynen) [inline variable/string.cpp:72]: Why are those listed twice?

**Review by jl-wynen — APPROVED** [review]
It is hard for me to say anything productive here because I missed the walkthrough. But it looks fine.

File: variable/string.cpp (hunk 6 above)
Comment (SimonHeybrock) [reply variable/string.cpp]: Bad merge

File: variable/bin_detail.cpp (hunk 5 above)
Comment (SimonHeybrock) [reply variable/bin_detail.cpp:27]: Nice catch! It did not actually get here, but that was due to inconsistent behavior of slicing `DataArray` or `Dataset` with length=0: Bin-edges were preserved as length=1 (since edge coords are always longer by one), but in this special case we actually need to set them to length=0 as well. See updates.

**Review by nvaytet — COMMENTED** [review]
I don't have much to add. After discussions on slack, changes make sense.

File: core/subbin_sizes.cpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+SubbinSizes &SubbinSizes::operator-=(const SubbinSizes &other) {
+  return *this = *this - other;
+}
+
+SubbinSizes SubbinSizes::cumsum_exclusive() const {
+  auto out = sizes();
+  scipp::index sum = 0;
+  for (auto &x : out) {
+    sum += x;
+    x = sum - x;
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/subbin_sizes.cpp:49]: I think we could write `element::exclusive_scan(sum, x)`, yes. Do you think that would be clearer?

File: dataset/test/bin_test.cpp [hunk 8]
```diff
@@ -314,8 +330,37 @@ TEST_P(BinTest, bin_by_group) {
...
+  auto edges = makeVariable<double>(Dims{Dim("group")}, Shape{6},
+                                    Values{-2, -1, 0, 1, 2, 3});
+  EXPECT_EQ(bin(binned, {edges}).data(), binned.data());
+
+  // Fewer bins than groups
+  edges = makeVariable<double>(Dims{Dim("group")}, Shape{3}, Values{-2, 0, 2});
+  EXPECT_NO_THROW(bin(binned, {edges}));
+}
+
+TEST_P(BinTest, rebin_various_edges_1d) {
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply dataset/test/bin_test.cpp:344]: Yes, basically, I could not think of a good way to check, and this was mainly intdended to cover many edge cases in the setup of bin overlaps. Since the parts doing the actual binning are unchanged the other tests hopefull cover those lines of code.

File: dataset/bin.cpp [hunk 9]
```diff
@@ -452,30 +511,25 @@ DataArray groupby_concat_bins(const DataArrayConstView &array,
...
-                      [](const scipp::index s) { return s == 1; })) {
-        throw except::BucketError("Not enough bin edges in dim " +
-                                  std::to_string(i) + ". Need at least 2.");
-      }
-    }
+  if (edges.empty() && groups.empty())
+    throw except::BucketError("Arguments 'edges' and 'groups' of scipp.bin are "
+                              "both empty. At least one must be set.");
+  for (const auto &edge : edges) {
+    const auto dim = edge.dims().inner();
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply dataset/bin.cpp:526]: There is a prior call to a helper function `validate_bin_args` that does the validation. Also, `inner()` returns `Dim::Invalid` if it is 0-D, which would cause exceptions down the line, so this should be safe regardless.

File: core/subbin_sizes.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  }
+  return {offset(), std::move(out)};
+}
+
+scipp::index SubbinSizes::sum() const {
+  return std::accumulate(sizes().begin(), sizes().end(), scipp::index{0});
+}
+
+void SubbinSizes::trim_to(const SubbinSizes &other) {
+  auto out = other; // TODO in principle this allocation could be avoided
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/subbin_sizes.cpp:59]: Complicated code.

File: dataset/test/bin_test.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (nvaytet) [reply dataset/test/bin_test.cpp:344]: Maybe just add a comment with this explanation?

File: core/subbin_sizes.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (nvaytet) [reply core/subbin_sizes.cpp:49]: It's also mainly to avoid duplication

File: dataset/test/bin_test.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (jl-wynen) [reply dataset/test/bin_test.cpp:344]: You could use `EXPECT_NOTHROW`

**Comment by SimonHeybrock** [comment]
CI passed here: https://dev.azure.com/scipp/scipp/_build/results?buildId=2838&view=results
